### PR TITLE
Fortran

### PR DESCRIPTION
**Pull request by Xiaoxuan1121**
In log.f90, the loop should occur from Line 490 to Line 514. Violation is 1 for the whole time and the printout on Line 491 shows that there is one entry is always larger than eps and never decreases.

### CODE REVIEW

**Comment by dajmcdon** [comment]
Which entry is always larger? The first one?

**Comment by Xiaoxuan1121** [comment]
> Which entry is always larger? The first one?

No, actually intercept doesn't change a bit during the violation check, it's resulting from one of the coefficients with magnitude e^-5 from Line 491 printout. I used the following test example for checking:

set.seed(123)
n <- 100
p <- 20
X <- matrix(data = rnorm(n*p, mean = 2, sd = 5), nrow = n, ncol = p)
beta_star <- c(rep(5, 5), c(5, -5, 2, 0, 0), rep(-5, 5), rep(0, 5))
groups <- rep(1:(p / 5), each = 5)
pr <- 1 / (1 + exp(-X %*% beta_star))
y0 <- rbinom(n, 1, pr)
y1 <- y0
y1[which(y1 == 0)] <- -1

fit_sparsegl <- sparsegl(X, y1, group = groups, loss = "logit", standardize = FALSE, asparse = 0, intercept = FALSE, nlam = 10)
fit_sparsegl <- sparsegl(X, y1, group = groups, loss = "logit", standardize = FALSE, asparse = 0, intercept = TRUE, nlam = 10)

**Comment by dajmcdon** [comment]
One possibility is that this is actually a convergence issue. The probabilities are nearly 0/1 always. I tried lowering maxit, but it still loops. I think it only counts outer loop iterations instead of inner loop iterations too. Are the fitted values also approaching the boundary? We made need to make a few modifications.

**Comment by Xiaoxuan1121** [comment]
> One possibility is that this is actually a convergence issue. The probabilities are nearly 0/1 always. I tried lowering maxit, but it still loops. I think it only counts outer loop iterations instead of inner loop iterations too. Are the fitted values also approaching the boundary? We made need to make a few modifications.

I couldn't output the fitted values because I was stuck at where producing the coefficients. It's really strange that all of a sudden this algorithm doesn't work any longer once I added an intercept to compute.

**Comment by dajmcdon** [comment]
The code you shared avoids the infinite loop when `intercept = FALSE`? I didn't try that one.

**Comment by Xiaoxuan1121** [comment]
> The code you shared avoids the infinite loop when `intercept = FALSE`? I didn't try that one.

I tried both actually when 'intercept = FALSE/TRUE' and both of them couldn't avoid the infinite loop. So 'b(0) = 0'  if 'intercept = FALSE' and 'b(0)' is fixed at some number in this infinite loop if `intercept = TRUE`.

**Comment by dajmcdon** [comment]
I think there are potentially a number of issues. The tests on this branch are failing. Even the `loss='ls'` option seems wrong. The `no-penalty` test seems to fail. I have a feeling some other things are broken here.

**Comment by Xiaoxuan1121** [comment]
> I think there are potentially a number of issues. The tests on this branch are failing. Even the `loss='ls'` option seems wrong. The `no-penalty` test seems to fail. I have a feeling some other things are broken here.

The no-penalty test indeed passed in main branch before, but after correcting the code in model.R:
on Line 28: changes colSum() to colMeans() inside if loop checking `standardize`, "sx <- sqrt(Matrix::colMeans(x^2))"
on Line 83: change * to /, "outlist$beta <- outlist$beta / xs".

if you change them back, then no-penalty test would work again.

**Comment by dajmcdon** [comment]
Right. So in adding logit, ls loss got broken. And we don't know if any of that propagated through to the issue. I reverted back a few commits to when logit was "working" and that's when I noticed that ls was broken. To thoroughly debug something, it's best to start with something that works!
